From e874d1a43544fe3572a2d3a8e6714a324302556d Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:13:58 -0400
Subject: [alsa] usb-audio: Add support for Focusrite Saffire 6 USB

Message-id: <140972823573-36-patchset-jkysela@redhat.com>
Patchwork-id: 90736
O-Subject: [RHEL 7.1 PATCH 035/442] ALSA: usb-audio: Add support for Focusrite Saffire 6 USB
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

Signed-off-by: Eduard Gilmutdinov <edgilmutdinov@gmail.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 11e424e88bd493b5d55d73d96c82bd889002ef30)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/usb/quirks-table.h b/sound/usb/quirks-table.h
index f5f0595..3454780 100644
--- a/sound/usb/quirks-table.h
+++ b/sound/usb/quirks-table.h
@@ -2569,6 +2569,46 @@ YAMAHA_DEVICE(0x7010, "UB99"),
   .type = QUIRK_MIDI_NOVATION
  }
 },
+{
+ USB_DEVICE(0x1235, 0x0010),
+ .driver_info = (unsigned long) &(const struct snd_usb_audio_quirk) {
+  .vendor_name = "Focusrite",
+  .product_name = "Saffire 6 USB",
+  .ifnum = QUIRK_ANY_INTERFACE,
+  .type = QUIRK_COMPOSITE,
+  .data = (const struct snd_usb_audio_quirk[]) {
+   {
+    .ifnum = 0,
+    .type = QUIRK_AUDIO_FIXED_ENDPOINT,
+    .data = &(const struct audioformat) {
+     .formats = SNDRV_PCM_FMTBIT_S24_3LE,
+     .channels = 4,
+     .iface = 0,
+     .altsetting = 1,
+     .altset_idx = 1,
+     .attributes = UAC_EP_CS_ATTR_SAMPLE_RATE,
+     .endpoint = 0x01,
+     .ep_attr = USB_ENDPOINT_XFER_ISOC,
+     .rates = SNDRV_PCM_RATE_44100 |
+       SNDRV_PCM_RATE_48000,
+     .rate_min = 44100,
+     .rate_max = 48000,
+     .nr_rates = 2,
+     .rate_table = (unsigned int[]) {
+      44100, 48000
+     }
+    }
+   },
+   {
+    .ifnum = 1,
+    .type = QUIRK_MIDI_RAW_BYTES
+   },
+   {
+    .ifnum = -1
+   }
+  }
+ }
+},
 
 /* Access Music devices */
 {
diff --git a/sound/usb/quirks.c b/sound/usb/quirks.c
index ede7536..8973070 100644
--- a/sound/usb/quirks.c
+++ b/sound/usb/quirks.c
@@ -674,8 +674,9 @@ static int snd_usb_gamecon780_boot_quirk(struct usb_device *dev)
 
 /*
  * Novation Twitch DJ controller
+ * Focusrite Novation Saffire 6 USB audio card
  */
-static int snd_usb_twitch_boot_quirk(struct usb_device *dev)
+static int snd_usb_novation_boot_quirk(struct usb_device *dev)
 {
  /* preemptively set up the device because otherwise the
   * raw MIDI endpoints are not active */
@@ -984,9 +985,9 @@ int snd_usb_apply_boot_quirk(struct usb_device *dev,
   /* Digidesign Mbox 2 */
   return snd_usb_mbox2_boot_quirk(dev);
 
- case USB_ID(0x1235, 0x0018):
-  /* Focusrite Novation Twitch */
-  return snd_usb_twitch_boot_quirk(dev);
+ case USB_ID(0x1235, 0x0010): /* Focusrite Novation Saffire 6 USB */
+ case USB_ID(0x1235, 0x0018): /* Focusrite Novation Twitch */
+  return snd_usb_novation_boot_quirk(dev);
 
  case USB_ID(0x133e, 0x0815):
   /* Access Music VirusTI Desktop */
-- 
1.7.1